README.md
# react-device-storage

An unified package for saving, reading and deleting data from/to sessionStorage, localStorage or cookies.
If user's device doesn't support sessionStorage and/or localStorage, the package uses cookies as a fallback.
Note it's designed for browsers only.

## Installation

`npm install --save react-device-storage`

## API

### cookies()

Enables using of cookies.

Returns `this`, so you can manipulate with data directly.

### localStorage()

Enables using of localStorage.

Returns `this`, so you can manipulate with data directly.

### sessionStorage()

Enables using of sessionStorage.

Returns `this`, so you can manipulate with data directly.

### save(key, value)

Saves data in selected storage. `key` should be a 'string', value may be whatever you want to store.

### read(key)

Reads data from selected storage.

### delete(key)

Deletes data from selected storage.

## Options

### cookieFallback
> Enable or disable cookies fallback **(boolean)**

### cookie
> Object of settings for cookies

#### cookie.path
> Cookie path<br />
> Use `/` as the path if you want your cookie to be accessible on all pages.<br />
> [default `/`]

#### cookie.maxAge
> Relative max age of the cookie from when the client receives it **(seconds)**
> [default `2592000` (30 days)]

#### cookie.domain
> Domain for the cookie<br />
> Use `https://*.yourdomain.com` if you want to access the cookie in all your subdomains.

#### cookie.secure
> Is only accessible through HTTPS? **(boolean)**

## Example

```js
import { Component } from 'react';
import DeviceStorage from 'react-device-storage';

import HelloWorld from './HelloWorld';

export default class MyApp extends Component {
  construct(props) {
    super(props);

    this.storage = new DeviceStorage({
      cookieFallback: true,
      cookie: {
        secure: true
      }
    }).localStorage();

    this.state = {
      name: this.storage.read('name')
    };
  }

  setName() {
    let name = 'User Name';

    this.storage.save('name', name);
    this.setState({
      name
    });
  }

  render() {
    return (
      <HelloWorld userName={this.state.name} onClick={this.setName.bind(this)} />
    );
  }
}
```

## Inspired by

- [react-cookie](https://www.npmjs.com/package/react-cookie)
- [react-localstorage](https://www.npmjs.com/package/react-localstorage)

## License

This code is released under the MIT license. Please see [LICENSE](https://github.com/asaritech/react-device-storage/blob/master/LICENSE) file for details.

## Contact

Reporting of any [issues](https://github.com/asaritech/react-device-storage/issues) are appreciated.

## Our projects

- Social Identity & Login Aggregator: [Ukey1](http://ukey.one)
- Ukey1 SDK for React: [ukey1-react-sdk](https://github.com/asaritech/ukey1-react-sdk/)


CHANGELOG.md
# CHANGELOG

Note that this package follows [SemVer](http://semver.org/).


## 1.x

- 1.0.0 (2017-02-14)
  - Initial release


webpack.config.js
var webpack = require('webpack');
var UglifyJsPlugin = webpack.optimize.UglifyJsPlugin;
var env = process.env.WEBPACK_ENV || 'dev';
var WebpackDevServer = require('webpack-dev-server');
var path = require('path');

var appName = 'react-device-storage';
var host = 'http://127.0.2.1';
var port = '9000';

var plugins = [], outputFile;

if (env === 'build') {
  plugins.push(new UglifyJsPlugin({ minimize: true }));
  outputFile = appName + '.min.js';
} else {
  outputFile = appName + '.js';
}

var config = {
  entry: './src/index.js',
  devtool: 'source-map',
  output: {
    path: __dirname + '/dist',
    filename: outputFile
  },
  module: {
    loaders: [
      {
        test: /(\.jsx|\.js)$/,
        loader: 'babel',
        exclude: /(node_modules|bower_components)/,
        query: {
          presets: ['react', 'es2015']
        }
      },
      {
        test: /(\.jsx|\.js)$/,
        loader: "eslint-loader",
        exclude: /node_modules/
      }
    ]
  },
  resolve: {
    root: path.resolve('./src'),
    extensions: ['', '.js', '.jsx']
  },
  plugins: plugins
};

if (env === 'dev') {
  new WebpackDevServer(webpack(config), {
    contentBase: './example',
    hot: true,
    debug: true
  }).listen(port, host, function (err, result) {
    if (err) {
      console.log(err);
    }
  });
  console.log('-------------------------');
  console.log('Local web server runs at http://' + host + ':' + port);
  console.log('-------------------------');
}

module.exports = config;


src/package.js
'use strict';

var cookie = require('cookie');
var lsExists = 'localStorage' in global;
var ssExists = 'sessionStorage' in global;

var keyPrefix = 'rds_react';

var vCookies = 1;
var vLocalStorage = 2;
var vSessionStorage = 3;

var raw = {};

function checkStorageVersion(v) {
  if (!v) {
    throw new Error('No storage version.');
  }
}

function DeviceStorage(o) {
  if (typeof o !== 'object') {
    o = {};
  }

  if (!(o.cookieFallback === true || o.cookieFallback === false)) {
    o.cookieFallback = true;
  }

  var cookie = {};

  if (typeof o.cookie === 'object') {
    if (typeof o.cookie.path === 'string') {
      cookie.path = o.cookie.path;
    } else {
      cookie.path = '/';
    }

    if (typeof o.cookie.maxAge === 'number') {
      cookie.maxAge = o.cookie.maxAge;
    } else {
      cookie.maxAge = 2592000; // 30 days
    }

    if (typeof o.cookie.domain === 'string') {
      cookie.domain = o.cookie.domain;
    }

    if (typeof o.cookie.secure === 'boolean') {
      cookie.secure = o.cookie.secure;
    }
  }

  this.o = o;
  this.o.cookie = cookie;
  this.v = 0;
  this.ls;
  this.ss;
}

DeviceStorage.prototype._storageVersion = function (v) {
  var testKey = [keyPrefix + 'test', 'test'].join('_');
  var errorMsg = 'Required storage version is not supported. You can enable cookieFallback.';

  try {
    if (process.browser) {
      if ((v === vLocalStorage && !lsExists) || (v === vSessionStorage && !ssExists)) {
        throw new Error(errorMsg);
      }

      if (v === vLocalStorage) {
        this.ls = global.localStorage;
        this.ls.setItem(testKey, 1);
        this.ls.removeItem(testKey);
      }

      if (v === vSessionStorage) {
        this.ss = global.sessionStorage;
        this.ss.setItem(testKey, 1);
        this.ss.removeItem(testKey);
      }
    } else {
      console.log('The package runs out of browser - simulation enabled.');
    }
  } catch (e) {
    if (this.o.cookieFallback) {
      v = vCookies;
    } else {
      throw new Error(errorMsg);
    }
  }

  return v;
}

DeviceStorage.prototype.cookies = function () {
  this.v = this._storageVersion(vCookies);
  return this;
};

DeviceStorage.prototype.localStorage = function () {
  this.v = this._storageVersion(vLocalStorage);
  return this;
};

DeviceStorage.prototype.sessionStorage = function () {
  this.v = this._storageVersion(vSessionStorage);
  return this;
};

DeviceStorage.prototype.save = function (key, value) {
  checkStorageVersion(this.v);
  var storageKey = [keyPrefix, key].join('_');
  var storageValue = JSON.stringify(value);

  if (process.browser) {
    try {
      switch (this.v) {
        case vLocalStorage:
          this.ls.setItem(storageKey, storageValue);
          break;
        case vSessionStorage:
          this.ss.setItem(storageKey, storageValue);
          break;
        case vCookies:
          document.cookie = cookie.serialize(storageKey, storageValue, this.o.cookie);
          break;
      }
    } catch (e) {
      throw new Error('Unable to save data.');
    }
  } else {
    raw[storageKey] = storageValue;
  }
}

DeviceStorage.prototype.read = function (key) {
  checkStorageVersion(this.v);
  var storageKey = [keyPrefix, key].join('_');
  var value, storageValue;
  var cookies;

  if (process.browser) {
    try {
      switch (this.v) {
        case vLocalStorage:
          storageValue = this.ls.getItem(storageKey);
          break;
        case vSessionStorage:
          storageValue = this.ss.getItem(storageKey);
          break;
        case vCookies:
          cookies = cookie.parse(document.cookie);
          storageValue = cookies && cookies[storageKey];
          break;
      }
    } catch (e) {
      throw new Error('Unable to read data.');
    }
  } else if ('storageKey' in raw) {
    storageValue = raw[storageKey];
  }

  if (storageValue) {
    try {
      value = JSON.parse(storageValue);
    } catch (e) {
      throw new Error('Unable to parse data.');
    }
  }

  return value;
}

DeviceStorage.prototype.delete = function (key) {
  checkStorageVersion(this.v);
  var storageKey = [keyPrefix, key].join('_');
  var cookieOptions = {};

  if (process.browser) {
    try {
      switch (this.v) {
        case vLocalStorage:
          this.ls.removeItem(storageKey);
          break;
        case vSessionStorage:
          this.ss.removeItem(storageKey);
          break;
        case vCookies:
          cookieOptions = this.o.cookie;
          cookieOptions.maxAge = 0;
          cookieOptions.expires = new Date(1970, 1, 1, 0, 0, 1);
          document.cookie = cookie.serialize(storageKey, '', cookieOptions);
          break;
      }
    } catch (e) {
      throw new Error('Unable to delete data.');
    }
  } else if ('storageKey' in raw) {
    delete raw[storageKey];
  }
}

module.exports = DeviceStorage;
